tests/NLog.UnitTests/LogToDiagnosticsTests.cs

```diff
@@ -0,0 +1,260 @@
+// 
+// * Neither the name of Jaroslaw Kowalski nor the names of its 
+//   contributors may be used to endorse or promote products derived from this
+//   software without specific prior written permission. 
+// 
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE 
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE 
+// ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE 
+// LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR 
+// CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+// SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS 
+// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN 
+// CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) 
+// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF 
+// THE POSSIBILITY OF SUCH DAMAGE.
+// 
+
+#define DEBUG
+
+#if !SILVERLIGHT && !__IOS__ && !__ANDROID__
+
+namespace NLog.UnitTests
+{
+    using NLog.Common;
+    using System;
+    using System.Collections.Generic;
+    using System.Diagnostics;
+    using System.Linq;
+    using System.Text;
+    using Xunit;
+    using Xunit.Extensions;
+
+    public class LogToDiagnosticsTests : NLogTestBase
+    {
+
+        [Theory]
+        [InlineData(null, null)]
+        [InlineData(false, null)]
+        [InlineData(null, false)]
+        [InlineData(false, false)]
+        public void ShouldNotLogInternalWhenLogToDiagnosticIsDisabled(bool? internalLogToDiagnostics, bool? logToDiagnostics)
+        {
+            var mockTraceListener = SetupTestConfiguration(LogLevel.Trace, CreateMockTraceListener, internalLogToDiagnostics, logToDiagnostics);
+
+            InternalLogger.Trace("Logger1 Hello");
+
+            Assert.Equal(0, mockTraceListener.Messages.Count);
+        }
+
+        /// <summary>
+        /// Helper method to setup tests configuration
+        /// </summary>
+        /// <param name="logLevel">The <see cref="LogLevel"/> for the log event.</param>
+        /// <param name="createTraceListenerFunc">Function which creates <see cref="TraceListener"/> instance.</param>
+        /// <param name="internalLogToDiagnostic">internalLogToDiagnostics XML attribute value. If <c>null</c> attribute is omitted.</param>
+        /// <param name="logToDiagnostic">Value of <see cref="InternalLogger.LogToDiagnostics"/> property. If <c>null</c> property is not set.</param>
+        /// <returns><see cref="TraceListener"/> instance.</returns>
+        private T SetupTestConfiguration<T>(LogLevel logLevel, Func<T> createTraceListenerFunc, bool? internalLogToDiagnostic, bool? logToDiagnostic) where T : TraceListener
+        {
+            var sb = new StringBuilder("<nlog ");
+            sb.AppendFormat("internalLogLevel='{0}'", logLevel);
```

Comment:
Using a stringbuilder is a good thing when combining strings, but in this case I would recommend 1 / multiple string.formats - it's more readable and the little performance  gain won't matter in unit tests